Support for qvm-template

WillyPillow opened the pull request:
Discussions: <https://groups.google.com/g/qubes-devel/c/PyJogqT1TUg>

Design doc: <https://gist.github.com/WillyPillow/b8a643ddbd9235a97bc187e6e44b16e4>

Major changes:

- Place appmenu whitelists in `qvm-features`
- Read `qvm-feature` and `qvm-pref` flags from `template.conf`. See document above for details
- Allow non-split root.img tars (may be removed in the future)
- Introduce `qvm-template` tool

WillyPillow committed:
Initial support for qvm-template.

WillyPillow committed:
Separate whitelist entries with spaces instead of newlines.

WillyPillow committed:
Verify values of boolean flags in template config.

WillyPillow committed:
Allow virt_mode other than pv.

WillyPillow committed:
Change "whitelist" to "menu-items" in qvm-features for clarity.

WillyPillow committed:
Initial commit of qvm-template.

Refer to <https://gist.github.com/WillyPillow/61ee5f48b7c5b7cc90c9fd2ec5c1b20d>
for previous revisions.

WillyPillow committed:
Better way of detecting VM.

marmarek commented on qubesadmin/tools/qvm_template.py:
```diff
@@ -0,0 +1,409 @@
+
+def do_list(args, app):
+    # TODO: Check local template name actually matches to account for renames
+    # TODO: Also display repo like `dnf list`
```
And also template state (for example if currently installed).

marmarek commented on qubesadmin/tools/qvm_template.py:
```diff
@@ -0,0 +1,409 @@
+                tpl_list.append((vm.name, version_str))
+
+    if args.available or args.all:
+        query_res = qrexec_repoquery(args, app)
```
You can call `qrexec_repoquery` just once and then reuse result in each case below.

marmarek commented on qubesadmin/tools/qvm_template.py:
```diff
@@ -0,0 +1,409 @@
+def build_version_str(evr):
+    return '%s:%s-%s' % evr
+
+def pretty_print_table(table):
```
There is already `qubesadmin.tools.print_table()` for this.

marmarek commented on qubesadmin/tools/qvm_template.py:
```diff
@@ -0,0 +1,409 @@
+            if not c:
+                break
+            # Write raw byte w/o decoding
+            sys.stdout.buffer.write(c)
```
This can be bad - terminal escape sequences can be dangerous. At lease filter out ESC character (either drop, or replace with something innocent), but to be on the safe side, I would do the same with anything outside of ASCII (so char code >128).

WillyPillow replied:
The reason for this and the byte-by-byte reads below are because I'm currently reusing curl's progress indicator. That being said, this is indeed an issue for security. Maybe I should either implement my own progress indicator or pull in pv/tqdm as a dependency?

WillyPillow replied:
Implemented a simple progress bar for the time being.

marmarek commented on qubesadmin/tools/qvm_template.py:
```diff
@@ -0,0 +1,409 @@
+        proc.stdin.write(payload.encode('UTF-8'))
+        proc.stdin.close()
+        while True:
+            c = proc.stderr.read(1)
```
Why doing it byte-by-byte?

marmarek commented on qubesadmin/tools/qvm_template.py:
```diff
@@ -0,0 +1,409 @@
+            tpl.features['template-name'] = name
+            # TODO: Store source repo
+
+def qrexec_popen(args, app, service, stdout=subprocess.PIPE, encoding='UTF-8'):
```
There is already `app.domains[...vmname...].run_service()` doing very similar thing.

Also, avoid decoding untrusted data (coming from another VM), at least not before some sanity checks, even if it's only UTF-8. Either use the simplest possible encoding (ascii), or drop automatic decoding entirely (preferred) - and then only decode parts that are actually needed (but still preferable with simple ascii).

WillyPillow replied:
Ahh, I initially did notice the function in the docs, but did not realize that it accepts a kwargs argument (as I needed to redirect the output to a file).

As for decoding, the only thing that needs to be decoded should be the output of TemplateSearch, which *should* indeed only require ASCII.

marmarek commented on qubesadmin/tools/qvm_template.py:
```diff
@@ -0,0 +1,409 @@
+    for r in args.repoid if args.repoid else []:
+        payload += '--repoid=%s\n' % r
+    payload += '--releasever=%s\n' % args.releasever
+    payload += spec + '\n'
```
All of the above could use a check if no unexpected characters are included (spaces, newlines). Better report a clear error early, than some confusing message when the command was misinterpreted.

WillyPillow replied:
Shouldn't whitespace characters (besides newlines) work as intended as long as the quoting is done correctly in the qrexec service?

marmarek replied:
Yes, you're right.

marmarek commented on qubesadmin/tools/qvm_template.py:
```diff
@@ -0,0 +1,409 @@
+            rpm_list.append(os.path.join(args.cachedir, target_file))
+    dl_list = dl_list_copy
+
+    download(args, app, path_override=args.cachedir, dl_list=dl_list)
```
To avoid accidental use of the file before verification, consider downloading to a differently-named file (like, append `.unverified` suffix or similar) and only rename after verification.

marmarek commented on qubesadmin/tools/qvm_template.py:
```diff
@@ -0,0 +1,409 @@
+            extract_rpm(name, rpmfile, target)
+            cmdline = [
+                'qvm-template-postprocess',
+                '--keep-source',
```
Why `--keep-source` (besides debugging)?

WillyPillow replied:
Oops, that was indeed something I did for debugging. As TemporaryDirectory removes the directory, I did not notice this while testing it afterwards.

marmarek commented on qubesadmin/tools/qvm_template.py:
```diff
@@ -0,0 +1,409 @@
+        return hdr
+
+def extract_rpm(name, path, target):
+    with open(path, 'rb') as in_file:
```
Why? You don't seem to use `in_file` here.

WillyPillow replied:
This seems to be the remains of the rpm2archive version. Will remove it later.

marmarek reviewed (changes requested):
Besides remarks below, take a look also at pylint output: https://travis-ci.com/github/QubesOS/qubes-core-admin-client/jobs/358989660#L350

Tests will need adjustments too, but that can be later step (when the actual tool will be done).

WillyPillow committed:
Various cleanup and improvements.

- `qvm-template list`: show template state
- `qvm-template list`: only call qubes.TemplateSearch once
- `qvm-template list`: use `qubesadmin.tools.print_table()` instead of own implementation
- `qvm-template download`: custom progress bar
- Use `run_service` instead of own implementation
- Remove some erroneous/redundant lines

marmarek commented on qubesadmin/tools/qvm_template.py:
```diff
@@ -0,0 +1,407 @@
+            # Clear previous bar
+            print(u'\u001b[1000D', end='', file=sys.stderr)
+            print(pct + '[' + ('#' * num_hash) + (' ' * num_space) + ']',
+                end='', file=sys.stderr)
```
Take a look at tqdm python module for progress bar. New (not released yet) pdf converter uses it: https://github.com/QubesOS/qubes-app-linux-pdf-converter/blob/master/qubespdfconverter/client.py

WillyPillow replied:
Ahh, so tqdm is actually available :) Mentioned it above but was not sure if it's okay to pull in a dependency for this.

WillyPillow committed:
Use tqdm for progress bar.

WillyPillow committed:
Check for newlines in qrexec arguments & improve error handling.

WillyPillow committed:
Add suffix for unverified RPMs.

WillyPillow committed:
Fix pylint warnings.

WillyPillow committed:
Support for {reinstall,downgrade,upgrade} operations.

Requires QubesOS/qubes-issues#5946 to be resolved.

WillyPillow committed:
Allow <package-spec>-like arguments for the list operation.

WillyPillow committed:
Initial implementation for "qvm-template info".

WillyPillow committed:
Initial implementation for "qvm-template search".

WillyPillow committed:
Add lock-file functionality for qvm-template install.

WillyPillow committed:
Print error messages if qubes.TemplateSearch fails.

WillyPillow committed:
Revamp "qvm-template search" and finish TODOs.

WillyPillow committed:
Check number of fields for qubes.TemplateSearch output.

WillyPillow committed:
qvm-template: Incorporate additional metadata in qubes.TemplateSearch.

WillyPillow committed:
qvm-template: Remove downloaded file if the download is interrupted.

WillyPillow committed:
qvm-template: Add --refresh option and allow DNF cache to be used.

WillyPillow committed:
qvm-template: Make pylint happy by changing "license" to "licence".

WillyPillow committed:
qvm-template*: Add option to specify pool to store created VM.

marmarek commented on qubesadmin/tools/qvm_template.py:
```diff
@@ -0,0 +1,802 @@
+    name_re = re.compile(r'^[A-Za-z0-9._+\-]*$')
+    evr_re = re.compile(r'^[A-Za-z0-9._+~]*$')
+    date_re = re.compile(r'^\d+-\d+-\d+ \d+:\d+$')
+    licence_re = re.compile(r'^[A-Za-z0-9._+\-()]*$')
```
`-` must be the last in the [] set, otherwise it is interpreted as a chars range.

WillyPillow replied:
Both in my testing and from regex101, the backslash before the `-` seem to work as intended?

marmarek replied:
Ah, indeed, I haven't noticed backslash.

marmarek commented on qubesadmin/tools/qvm_template.py:
```diff
@@ -0,0 +1,802 @@
+                continue
+
+            result.append((name, epoch, version, release, reponame, dlsize,
+                buildtime, licence, url, summary, description))
```
Perhaps use `collections.namedtuple()`? Will ease handling this many items.

marmarek commented on qubesadmin/tools/qvm_template.py:
```diff
@@ -0,0 +1,802 @@
+
+    def append_info(data, status, install_time=None):
+        name, epoch, version, release, reponame, dlsize, \
+            buildtime, licence, url, summary, description = data
```
With namedtuple you could access `data.name` etc below, without this unpacking.

WillyPillow committed:
qvm-template: {reinstall,{up,down}grade}: Better handling and checks for existing version.

WillyPillow committed:
qvm-template: Switch to namedtuples and other slight cleanup.

WillyPillow committed:
qvm-template: Reorder functions.

WillyPillow committed:
qvm-template: Make pylint happy.

WillyPillow committed:
qvm-template: Check that template spec is not "---".

WillyPillow committed:
qvm-template: Fix typo.

WillyPillow committed:
qvm-template: Check that template is managed by qvm-template before accessing relevant features.

marmarek commented on qubesadmin/tools/qvm_template.py:
```diff
@@ -0,0 +1,831 @@
+
+def is_managed_template(vm):
+    return 'template-name' in vm.features \
+        and vm.name == vm.features['template-name']
```
This can be simplified as `vm.features.get('template-name', None) == vm.name`. A not so obvious side effect is that, it avoids listing vm features, which may save on required Admin API permissions.

marmarek commented on qubesadmin/tools/qvm_template.py:
```diff
@@ -0,0 +1,831 @@
+                " argument should not contain '\\n'.")
+
+    payload = ''
+    for repo in args.enablerepo if args.enablerepo else []:
```
You can add `default=[]` to the `parser.add_argument()` call to simplify this (and other instances below).

marmarek commented on qubesadmin/tools/qvm_template.py:
```diff
@@ -0,0 +1,831 @@
+
+def qrexec_repoquery(args, app, spec='*', refresh=False):
+    proc = qrexec_popen(args, app, 'qubes.TemplateSearch')
+    payload = qrexec_payload(args, app, spec, refresh)
```
Better first prepare payload and only then open the qrexec call - this way if the `qrexec_payload` raises an exception, you saves the qrexec call.

WillyPillow committed:
qvm-template: Better args parsing: Use subparsers and complain about unknown args if the operation is not "remove".

WillyPillow committed:
qvm-template: Use repo file from qubes-repo-templates.

WillyPillow committed:
qvm-template: Use "vm.features.get" instead of explicit membership check.

WillyPillow committed:
qvm-template: Defer qrexec calls so that they can be omitted if exceptions are raised.

WillyPillow committed:
qvm-template: Partially include docstrings and type hints.

WillyPillow committed:
qvm-template: More docstrings.

WillyPillow committed:
qvm-template: Initial support for machine-readable listings.

WillyPillow committed:
qvm-template: Include module in documentation.

WillyPillow committed:
qvm-template: Tidy up code responsible for output in {info,list} operations.

WillyPillow committed:
qvm-template: Replace newlines in machine-readable output.

WillyPillow committed:
qvm-template: Initial implementation of repolist.

WillyPillow committed:
qvm-template: Remove default 'repo_files' entry if other entries have been specified by the user.

WillyPillow committed:
qvm-template: Fix missing args for install operations.

WillyPillow committed:
qvm-template: Add confirmation for dangerous operations; verify signatures once instead of twice by returning header after verification.

WillyPillow committed:
qvm-template-postprocess: Make pylint happy.

WillyPillow committed:
qvm-template: Fix type hints.

marmarek commented on qubesadmin/tools/qvm_template.py:
```diff
@@ -0,0 +1,1269 @@
+            if not package_hdr:
+                parser.error('Package \'%s\' verification failed.' % rpmfile)
+
+            if reponame != '@commandline':
```
`path != rpmfile` would be clearer.

marmarek commented on qubesadmin/tools/qvm_template.py:
```diff
@@ -0,0 +1,1269 @@
+            package_name = package_hdr[rpm.RPMTAG_NAME]
+            if not package_name.startswith(PACKAGE_NAME_PREFIX):
+                parser.error(
+                    'Illegal package name for package \'%s\'.' % rpmfile)
```
Shouldn't this be before rename?

marmarek commented on qubesadmin/tools/qvm_template.py:
```diff
@@ -0,0 +1,1269 @@
+            confirm = ''
+            while confirm != 'y':
+                confirm = input('Are you sure? [y/N] ').lower()
+                if confirm == 'n':
```
This definitely should be `confirm != 'y'`.

WillyPillow replied:
The idea was to ask again if the input is neither 'y' or 'n', hence the while loop. AFAIK some CLI tools do this.

(I'm also okay with making it ask only once for simplicity though.)

WillyPillow committed:
qvm-template: Slight improvements to package verification.

WillyPillow committed:
qvm-template: Add purge operation.

WillyPillow committed:
qvm-template: Add option to specify RPM keyring location.

WillyPillow committed:
qvm-template: Add support for JSON output.

WillyPillow committed:
qvm-template: Add option to disable download progress bar.

marmarek commented on qubesadmin/tools/qvm_template.py:
```diff
@@ -0,0 +1,1387 @@
+            dummy = '%s-%d' % (orig_dummy, cnt)
+            cnt += 1
+        if dummy not in app.domains:
+            dummy_vm = app.add_new_vm('TemplateVM', dummy, 'red')
```
Above you look at `template-dummy` feature, shouldn't you also add it here?

WillyPillow committed:
qvm-template: Make sure that template-dummy is set and used properly.

WillyPillow committed:
qvm-template: Initial manpage.

WillyPillow committed:
qvm-template: Tweak machine-readable output format.

marmarek commented:
Generally the changes looks good now. Take a look at test failure (travis logs) - you need to adjust few tests for new menu-items feature.

marmarta commented on qubesadmin/tools/qvm_template.py:
```diff
@@ -0,0 +1,1396 @@
+        with open('/usr/share/qubes/marker-vm', 'r') as fd:
+            # Get last line (in the format `x.x`)
+            return fd.readlines()[-1].strip()
+    return subprocess.check_output(['lsb_release', '-sr'],
```
Either this is missing a dependency (there's no lsb_release in dom0 by default), or maybe you could just directly read /etc/os-release ? it seems a bit overkill to use a separate process and program to parse a simple file - especially as redhat-lsb-core has several otherwise useless dependencies.

WillyPillow replied:
Replaced with reading /etc/os-release.

(Did not realize that there's no lsb_release by default -- upon inspection it seemed to be pulled in by `tlp` on my box.)

WillyPillow committed:
qvm-template: Eliminate use of lsb_release

WillyPillow committed:
qvm-template: Replace "template-install-time" with "template-installtime" for consistency

WillyPillow committed:
qvm-template: Improve help message for --upgrades

WillyPillow committed:
qvm-template: Fix qvm_template_postprocess tests

WillyPillow committed:
Merge remote-tracking branch 'origin/master' into qvm-template

WillyPillow committed:
Fix CI dependencies

WillyPillow committed:
qvm-template: Make pylint happy

marmarek commented:
> The diff coverage is `2.36%`.

This could use some tests...

WillyPillow commented:
> > The diff coverage is `2.36%`.
> 
> This could use some tests...

Yes, I'm working on it right now :)

Other than that, CI seems to pass now.

That being said, I also had to ignore the modules `rpm` and `dnf` for `pylint` as they aren't on PyPI. Is there a better way of approaching this?
(For `rpm`, it should possible to install it as a system package. OTOH, `dnf` doesn't seem to be packaged on Ubuntu.)

WillyPillow commented:
Side note: I'm unsure why `policy/qubesos/code-signing` gives `Unable to verify (no valid key found)`.

[My (sub)key](http://hkps.pool.sks-keyservers.net/pks/lookup?op=vindex&fingerprint=on&search=0x1C16C70EE7C31C84) does seem to be on `hkps.pool.sks-keyservers.net`.

IIRC, the SKS keyservers are kind of unstable recently. Don't know if it's because of that. (Running `--recv-keys` on my end also seems to fail most of the time.)

marmarek commented:
> IIRC, the SKS keyservers are kind of unstable recently. Don't know if it's because of that.

Yes, that's exactly it...


> That being said, I also had to ignore the modules `rpm` and `dnf` for `pylint` as they aren't on PyPI. Is there a better way of approaching this?

You'll face this issue for tests too. As a workaround you can create stubs in test-packages directory. For tests, you'd most likely mock them anyway.

WillyPillow committed:
Add stubs for rpm module and initial tests for qvm-template install

WillyPillow committed:
qvm-template: Update docs for --updatevm

WillyPillow commented:
CI is failing because of timezones. Will fix it later.

WillyPillow committed:
qvm-template: Fix timezone issues by storing timezone explictly in features

WillyPillow committed:
qvm-template: Fix compatibility with Python 3.6

WillyPillow committed:
qvm-template: Fix compatibility with Python 3.6

WillyPillow committed:
Revert "qvm-template: Fix compatibility with Python 3.6"

This reverts commit 199996e7b8757b6872b5064d39527fd3491d8e94.

WillyPillow commented:
It appears that `datetime.fromisoformat` does not exist in Python 3.6...

Considering <https://groups.google.com/g/qubes-devel/c/f7YcmR6Irus/m/dEgLk92nAgAJ>, should I, e.g., store instead the unix timestamp in feature tags, or just give up support for 3.6?

(On further reflection, it's probably simpler to stick to a fixed date format like `YYYY-MM-DD HH:mm:ss`. Also easier to define precisely in the docs compared to the current solution.)

WillyPillow committed:
qvm-template: Fix date formats to "%Y-%m-%d %H:%M:%S"

WillyPillow committed:
qvm-template: Add tests for qrexec_payload

WillyPillow committed:
Make TestProcess.communicate return str instead of IO object

WillyPillow committed:
qvm-template: Add partial tests for qrexec_repoquery

WillyPillow committed:
qvm-template: Fix compatibility with Python 3.6 in tests

WillyPillow committed:
qvm-template: Make pylint happy

marmarek commented:
> (On further reflection, it's probably simpler to stick to a fixed date format like `YYYY-MM-DD HH:mm:ss`. Also easier to define precisely in the docs compared to the current solution.)

Yes, this is fine. But please either include also timezone info in the format, or always store in UTC.

marmarek commented:
The qvm-template tool itself I think looks pretty good already. But take a look at the [test coverage report](https://codecov.io/gh/QubesOS/qubes-core-admin-client/pull/145/diff?src=pr&el=tree#diff-cXViZXNhZG1pbi90b29scy9xdm1fdGVtcGxhdGUucHk=) - some functions are missing tests. The trivial ones are that important, but for example get_dl_list or search could use some tests.

WillyPillow commented:
> Yes, this is fine. But please either include also timezone info in the format, or always store in UTC.

Yes, I think the times are all in UTC now.

(The times returned by qubes.TemplateSearch are also UTC, cf. <https://github.com/rpm-software-management/dnf/blob/master/dnf/cli/commands/repoquery.py#L711>.)

> The qvm-template tool itself I think looks pretty good already. But take a look at the [test coverage report](https://codecov.io/gh/QubesOS/qubes-core-admin-client/pull/145/diff?src=pr&el=tree#diff-cXViZXNhZG1pbi90b29scy9xdm1fdGVtcGxhdGUucHk=) - some functions are missing tests. The trivial ones are that important, but for example get_dl_list or search could use some tests.

I see. I was planning to do it in a bottom-up order, hence the current set of tests. Regardless, the basic functions should be mostly done, and right now I'm working on the more complicated operations like the ones you mentioned.

WillyPillow committed:
qvm-template: Add tests for qrexec_repoquery and get_dl_list

WillyPillow committed:
qvm-template: Fix minor bugs

* Incomprehensive spec filtering in `list_templates`
* Type error of `install_time` in `list_templates`
* Incorrect version comparision in `search`

WillyPillow committed:
qvm-template: Add tests for functions list_templates and search

WillyPillow committed:
qvm-template: Mock print_table in tests for consistent output

WillyPillow committed:
qvm-template: Fix broken indention

marmarek commented:
I reviewed the updated version and I see you also found some minor bugs while writing tests :) 
Looking at test coverage, those are non-trivial parts not covered with tests:
 - "download" function (especially adding/dropping .rpm suffix)
 - install with override_existing=True
 - "remove" function
 - qvm-template-postprocess - processing template.conf

WillyPillow committed:
qvm-template: Add tests for download function and fix minor bugs

marmarek commented:
Let me know when you think it is ready for merge and also remove "WIP" tag from the title.

marmarek commented:
PipelineRetry

marmarek committed:
qvm-template: use UpdateVM by default

Do not hardcode sys-firewall

marmarek committed:
qvm-template: call rpmkeys --checksig for signature verification

RPM API is confusing and it's easy to get it wrong when verifying
package signatures.
Call 'rpmkeys --checksig' which is more rebust here - RPM authors should
know how to use their API.

QubesOS/qubes-issues#2534

marmarek committed:
qvm-template: download templates to a temporary directory

Avoid risk of conflicting downloads to the same directory, reusing
partial downloads, leaving broken files etc. Move template package out
of temporary directory only after its verified.

QubesOS/qubes-issues#2534

marmarek committed:
qvm-template: handle template extraction failure

QubesOS/qubes-issues#2534

marmarek committed:
qvm-template: add copyright header

marmarek committed:
qvm-template: minor improvements

- rename parser_gen to get_parser - for consistency with other tools
- clarify 'storage pool'
- move '-' to the end in regex characters list

QubesOS/qubes-issues#2534

marmarek committed:
deb,rpm: depend on qubes-repo-templates

For package repos definitions + keys.

marmarek committed:
qvm-template-postprocess: improve data validation

- validate if IP has correct syntax
- print warning if value is invalid

QubesOS/qubes-issues#2534

marmarek committed:
qvm-template: improve install lock

Use fcntl.flock() instead of just file existence check, so it won't fail
on a stale lock file. While at it, move locking to a function decorator,
to de-clutter the install function a bit. This will allow reducing
indentation level, but don't do it yet, to make the patch readable.

Move lock testing into a separate test, and remove it from install
tests.

QubesOS/qubes-issues#2534

marmarek committed:
qvm-template: cleanup install function

Remove now unused try/finally in install() and reduce indentation.

No functional change.

marmarek committed:
qvm-template: remove downloaded package after installation

At least by default. Otherwise they will pile up in the cache dir.

marmarek committed:
qvm-template-preprocess: remove confusing message

Don't confuse user during normal template install.

marmarek committed:
Store template repo configuration in /etc

This way it's easier to permanently enable repos.

marmarek committed:
qvm-template: use QubesArgumentParser

It produces consistent help for subcommands and already handles
--verbose/--quiet.

marmarek committed:
qvm-template: allow global arguments after action name

It's convenient to use for example
`qvm-template list --enablerepo=*-testing`
Previously, _some_ options needed to be before action name.

marmarek committed:
qvm-template: factor filter_version() out of get_dl_list()

This allows reusing version filtering (getting only a single version per
template) in other places.

For equal versions packages, prefer the one from non-testing repository.

marmarek committed:
qvm-template: by default list only latest available template

But add --all-versions option to get all the available versions.

marmarek committed:
qvm-template: use key specified in the repo definition if possible

This makes the package verified against _only_ the key specified in the
repo config, not all the trusted keys.
If repo does not specify a key, use the default one (change this to a
single file, instead of the whole directory). Existing 'gpgkey' entry
pointing at non-existing file will result in an error.

marmarek committed:
qvm-template-postprocess: do not generate appmenus twice

Skip initial generate, as it's done before actual menu entries are
extracted from the template. But do call it if we aren't going to
extract menu entries initially - it will create just "settings" menu
entry.

marmarek committed:
qvm-template: improve error reporting

Do not print the whole traceback by default - do that only when
--verbose is used.

marmarek committed:
qvm-template: verify template package signature directly at download

Make the download() function save the package into a temporary space and
move to the target location only after checking the signature. This is
safer option than requiring all callers to explicitly verify the
signature. Also, make the download() function verify if the template
name inside the package matches what was requested.
Especially, make `qvm-template download` action verify the signature
too.

On `qvm-template install` avoid checking the signature again for
downloaded packages, by passing extra argument to the verify_rpm()
function. But still verify signature of packages loaded from disk.

marmarek committed:
qvm-template: default confirm to 'n'

Capital 'N' in the prompt suggests it is the default - really make it
the default.

marmarek committed:
qvm-template-postprocess: treat missing appmenus files as warnings only

Do not fail if *-whitelisted-appmenus.list files are not included in the
template package, only log an error.

While at it, use pathlib there to make the code a bit nicer.

marmarek committed:
qvm-template-postprocess: extract config handling into separate function

Keep post_install() short.

marmarek committed:
qvm-template-postprocess: fix allowed features list

Add missing coma, otherwise 'gui-emulated' and 'qrexec' were glued
together.

marmarek committed:
tests: qvm-template-postprocess - template.conf handling

marmarek committed:
qvm-template: mute pylint complains about typing.NamedTuple

This is false positive, PyCQA/pylint#3732

marmarek committed:
tests: some more for qvm-template

QubesOS/qubes-issues#

WillyPillow committed:
tests: add tests for qvm-template remove

WillyPillow committed:
tests: fix tests for verify_rpm involving incorrect template names

WillyPillow committed:
qvm-template: test != 1 instead of == 0 for template-dummy feature

WillyPillow commented:
Thanks for the commits! Apologies for the delay -- honestly I just took up too much work this semester, got pretty burnt out in the end, and had to push things back a bit :(

WillyPillow reviewed (commented):
Soon after posting this I realized that I already _had_ something similar (`add_new_vm_side_effect`). That was pretty silly...
Edit: Accidentally deleted the comment that this was referring to when I was trying to close the review :(

WillyPillow committed:
tests: add more tests re. install, remove, and get_keys_for_repos

marmarek commented:
You can assign a list of results to `self.expected_calls[...]` - they will be returned in order for subsequent calls.

WillyPillow committed:
qvm-template: only ask for confirmation during install if something is being done

WillyPillow committed:
qvm-template: update comments to reflect e424c7d

WillyPillow committed:
tests: fix mock return values of get_dl_list when testing `qvm-template reinstall`

WillyPillow committed:
tests: add tests for qvm-template reinstall/up/downgrade when nothing needs to be done

WillyPillow committed:
tests: improve TestProcess behavior

- Have it actually write to the given stdout handle.
- Return the return code for `poll` instead of returning `None`, so that
the process is observed to terminate.

WillyPillow committed:
tests: add tests for other qvm-template functions

marmarek committed:
qvm-template: fix downloading template for install

Donwload a template into a cache dir, not into default of `qvm-template
download` (current directory).

marmarek reviewed (approved):
The code looks good now. There is a missing test for repolist, but it isn't very complex function and also not a critical functionality, so I'm ok with merging it as is.

Also I've been testing this whole feature manually for some time already and it works fine!

WillyPillow commented:
Thanks. IIRC, the main reason I skipped the tests for `repolist` is that the dnf dependency isn't available everywhere and writing a stub for it doesn't seem trivial.

Again, I really appreciate the help along the way!
